# metagenomics training

In this report you will find all the information necessary to follow the steps to run a basic metagenomics analysis.

## Training overview
During this training we will following these steps:
* [Mapping](#mapping): Introduction to mapping.
* [Assembly](#assembly): Introduction to assembly.
* [Register/Login](#register/login): Register or login into VirusTAP and Galaxy.
* [Data](#data): Upload data for the analysis.
* [Parameters/Steps](#parameters/steps): Select the desired parameters and steps for the pipeline.
* [Run/Results](#run/results): Execute and explore the results.
* [Kraken](#kraken): K-mer mapping approach using Galaxy.
* [Combined approach](#combinedapproach): Both mapping and assambly can be combined to better understand your sample.

## Mapping

Mapping is a type of alignment that consist in placing a sequence inside a larger sequence. For example, it is used to determine the position of a read or contig in a reference genome.

![mapping](https://software.broadinstitute.org/software/igv/sites/cancerinformatics.org.igv/files/images/del_multiple.jpg)

This approach is useful when you have a reference genome for the same species of the organisms you are working with, it is faster than a *de-novo* assambly, and also allows variant discovery.

## Assembly

Assembling consist in reconstructing a representation of the original sequenced genome from shorter fragments generated by the sequencer machine called reads. It does not need a reference genome, and it is very uncommon the whole genome is reconstructed; they usually return only contigs that have to be elongated into scaffolds. Between those contigs there will be gaps, portions of the genome that could not be rebuilt.

![assembling](https://i0.wp.com/thesequencingcenter.com/wp-content/uploads/2019/01/denovo_assembly.jpg?w=1000&ssl=1)

This approach takes longer to compute, but it can discover new genomes.

## Register/login

In order to access VirusTAP, you first will need to log in or create a new academic account at [VirusTAP](https://gph.niid.go.jp/virustap/system_in):

![virustap_login](../docs/images/virustap_login.png)

## Data

We will start by giving a name to our analysis and uploading the raw data. In our case, we are going to use the files `Rota_L001_R1_001.fastq.gz` and `Rota_L001_R2_001.fastq.gz`. For this server, the data has to be compressed in this specific format.

![virustap_data](../docs/images/virustap_data.png)

The data can be downloaded [here](https://we.tl/t-jMLNdSmn2K).

## Parameters/Steps

The next step is to select the right paremeters for our analysis in each step of the pipeline we are going to run.

![virustap_parameters](../docs/images/virustap_parameters.png)

As you can see, the pipeline contains the following steps:

- **Quality trimming and adapter removal**: The first step is the quality control. We will use `skewer` as trimming software and set the other values to trimm 5 at the 5' end, and everything at the 3' end that has lower than 14 quality. Then, ignore reads with less average quality than 15 and also shorter than 80bp.

- **Read subtraction**: Then, we are going to remove the host. In this case, that's human. We are also only going to look for viruses, so we are going to also check to remove all non-viral reads. The software we will use for this step is `bwasw`, which is a mapper that we will use for matching (and removing) reads against the selected reference genomes.

- **De novo assembly**: Now we are going to assemble the viral reads, by using `a5_miseq`. This will get us all the contigs of the different viruses in our sample.

- **Homology search / read mapping**: Finally, we are going to blast the contigs to discover what are they, and which species they seem to belong to.

Once we are finally happy with our analysis configuration, we can start the analysis:

![virustap_run](../docs/images/virustap_run.png)
